receiver/prometheusexecreceiver/receiver.go

```diff
@@ -0,0 +1,313 @@
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"net"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	sdconfig "github.com/prometheus/prometheus/discovery/config"
+	"github.com/prometheus/prometheus/discovery/targetgroup"
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/receiver/prometheusreceiver"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/prometheusexecreceiver/subprocessmanager"
+)
+
+const (
+	minPortRange       int           = 10000            // Minimum of the port generation range
+	maxPortRange       int           = 11000            // Maximum of the port generation range
+	portTemplate       string        = "{{port}}"       // Template for port in strings
+	healthyProcessTime time.Duration = 30 * time.Minute // healthyProcessTime is the default time a process needs to stay alive to be considered healthy
+	healthyCrashCount  int           = 3                // healthyCrashCount is the amount of times a process can crash (within the healthyProcessTime) before being considered unstable - it may be trying to find a port
+
+)
+
+// Local random seed to not override anything being used globally
+var random *rand.Rand = rand.New(rand.NewSource(time.Now().UnixNano()))
+
+type prometheusReceiverWrapper struct {
+	logger   *zap.Logger
+	config   *Config
+	consumer consumer.MetricsConsumerOld
+
+	// Prometheus receiver config
+	receiverConfig *prometheusreceiver.Config
+
+	// Subprocess data
+	subprocessConfig *subprocessmanager.SubprocessConfig
+
+	// Receiver data
+	originalPort       int
+	prometheusReceiver component.MetricsReceiver
+	context            context.Context
+	host               component.Host
+}
+
+// new returns a prometheusReceiverWrapper
+func new(logger *zap.Logger, config *Config, consumer consumer.MetricsConsumerOld) *prometheusReceiverWrapper {
+	return &prometheusReceiverWrapper{logger: logger, config: config, consumer: consumer}
+}
+
+// Start creates and calls the function that handles the prometheus_exec receiver
+func (wrapper *prometheusReceiverWrapper) Start(ctx context.Context, host component.Host) error {
+	factory := &prometheusreceiver.Factory{}
+
+	customName := getCustomName(wrapper.config)
+
+	subprocessConfig, ok := getSubprocessConfig(wrapper.config, customName)
+	if ok != nil {
+		return fmt.Errorf("unable to generate the subprocess config: %w", ok)
+	}
+
+	receiverConfig := getReceiverConfig(wrapper.config, customName)
+
+	receiver, ok := factory.CreateMetricsReceiver(wrapper.context, wrapper.logger, receiverConfig, wrapper.consumer)
+	if ok != nil {
+		return fmt.Errorf("unable to create Prometheus receiver: %w", ok)
+	}
+
+	wrapper.originalPort = wrapper.config.Port
+	wrapper.subprocessConfig = subprocessConfig
+	wrapper.receiverConfig = receiverConfig
+	wrapper.prometheusReceiver = receiver
+	wrapper.context = ctx
```

Comment:
It pretty unusual to store context. I don't think you really want to do this. It's a scoped bundle of attributes that is passed around, and if you store it you may end up using it in a different scope.

Generally when you start an asynchronous task as you do below, you want to create a new context for that goroutine.